### PR TITLE
#447: unit tests for LazyResolutionService

### PR DESCRIPTION
**Pull request by JudeNiroshan**
#### Summary

Unit tests for LazyResolutionService

#### Todo

- Tests
    - [x] Unit 
    - [x] Integration
- [x] Documentation
- [ ] Add Release Notes entry.

### CODE REVIEW

File: src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java [hunk 1]
```diff
@@ -60,6 +63,13 @@ public UnresolvedReferencesServiceImpl(@Nonnull final ProductSyncOptions baseSyn
     @Nonnull
     @Override
     public CompletionStage<Optional<WaitingToBeResolved>> save(@Nonnull final WaitingToBeResolved draft) {
+
+        if (isBlank(draft.getProductDraft().getKey())) {
+            syncOptions.applyErrorCallback(format(SAVE_FAILED, draft.getProductDraft().getKey(),
+                    "Draft key is blank!"));
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java]: 1. It's nice to have but we don't need it at the moment (https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) since we know we always call it with a non-blank key. So there is no need to add the logic and test it and so on. 
2. "Draft key is blank!" is not very helpful for the user in this case, some information in the draft like the value maybe could be useful for the user to know more information about this case.

**Comment by heshamMassoud** [comment]
Unit tests are also missing for the `WaitingToBeResolved` class.

File: src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java [hunk 2]
```diff
@@ -37,6 +39,7 @@ public UnresolvedReferencesServiceImpl(@Nonnull final ProductSyncOptions baseSyn
     @Nonnull
     @Override
     public CompletionStage<Set<WaitingToBeResolved>> fetch(@Nonnull final Set<String> keys) {
+        keys.remove(EMPTY);
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java]: - It's not a good practice to mutate in the user input.
- What is the purpose of removing `EMPTY` at all? We should not be expecting it to be passed.

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        assertThat(errorMessages).hasSize(1);
+        assertThat(errorExceptions).hasSize(1);
+        assertThat(errorMessages.get(0))
+                .isEqualTo("Failed to save CustomObject with key: 'null'. Reason: Draft key is blank!");
+    }
+
+    @Test
+    void fetch_WithEmptyKeySet_ShouldReturnEmptySet() {
+        Set<String> keys = new HashSet<>();
+        keys.add("");
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: why do you add it `""`, why not test with an originally empty set?

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        final Set<WaitingToBeResolved> beResolvedOptional = service
+                .fetch(keys).toCompletableFuture().join();
+
+        assertThat(beResolvedOptional).isEmpty();
+    }
+
+
+    @Test
+    void fetch_WithSuccessfulMockCtpResponse_ShouldReturnMock() {
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: what about fetch with unsuccessful?

Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: I find tests to read when the bodies are separated with comments:

```java
// preparation
...
// test
...
// assertion
...
```
What do you think about this?

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        final CustomObject mock = mock(CustomObject.class);
+        when(mock.getContainer()).thenReturn("commercetools-sync-java.UnresolvedReferencesService.productDrafts");
+
+        final ProductDraft productDraft = mock(ProductDraft.class);
+        when(productDraft.getKey()).thenReturn("product-draft-key");
+        final Set<String> missingRefs = new HashSet<>();
+        missingRefs.add("test-ref");
+        WaitingToBeResolved valueObj = new WaitingToBeResolved(productDraft, missingRefs);
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: can be simplified to:

```java
final WaitingToBeResolved waitingDraft = new WaitingToBeResolved(productDraft, singleton("test-ref"));
```

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        when(mock.getValue()).thenReturn(valueObj);
+
+        Set<String> keys = new HashSet<>();
+        keys.add("product-draft-key");
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: You can inline this directly on line 158 with just

```java
final Set<WaitingToBeResolved> toBeResolvedOptional = service.fetch(singleton("product-draft-key")).toCompletableFuture().join();
```

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                .fetch(keys).toCompletableFuture().join();
+
+        assertThat(beResolvedOptional).isEmpty();
+    }
+
+
+    @Test
+    void fetch_WithSuccessfulMockCtpResponse_ShouldReturnMock() {
+        final CustomObject mock = mock(CustomObject.class);
+        when(mock.getContainer()).thenReturn("commercetools-sync-java.UnresolvedReferencesService.productDrafts");
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: this is not needed, right?

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        missingRefs.add("test-ref");
+        WaitingToBeResolved valueObj = new WaitingToBeResolved(productDraft, missingRefs);
+        when(mock.getValue()).thenReturn(valueObj);
+
+        Set<String> keys = new HashSet<>();
+        keys.add("product-draft-key");
+
+        PagedQueryResult pagedQueryResult = mock(PagedQueryResult.class);
+        when(pagedQueryResult.getCount()).thenReturn(1L);
+        when(pagedQueryResult.getOffset()).thenReturn(0L);
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: is this needed?

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        when(productDraft.getKey()).thenReturn("product-draft-key");
+        final Set<String> missingRefs = new HashSet<>();
+        missingRefs.add("test-ref");
+        WaitingToBeResolved valueObj = new WaitingToBeResolved(productDraft, missingRefs);
+        when(mock.getValue()).thenReturn(valueObj);
+
+        Set<String> keys = new HashSet<>();
+        keys.add("product-draft-key");
+
+        PagedQueryResult pagedQueryResult = mock(PagedQueryResult.class);
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: `final`

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        when(productSyncOptions.getCtpClient().execute(any())).thenReturn(completedFuture(pagedQueryResult));
+
+        final Set<WaitingToBeResolved> toBeResolvedOptional = service.fetch(keys).toCompletableFuture().join();
+
+        assertThat(toBeResolvedOptional).isNotEmpty();
+        assertThat(toBeResolvedOptional).containsOnly(valueObj);
+    }
+
+    @Test
+    void delete_WithUnsuccessfulMockCtpResponse_ShouldReturnProperException() {
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: What about successful delete?

**Comment by lojzatran** [comment]
Where is `LazyResolutionService`? I don't see any class with this name.

**Comment by heshamMassoud** [comment]
> Where is `LazyResolutionService`? I don't see any class with this name.

@lojzatran `UnresolvedReferencesService` used to be `LazyResolutionService`

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                    errorMessages.add(errorMessage);
+                    errorExceptions.add(errorException);
+                })
+                .build();
+        service = new UnresolvedReferencesServiceImpl(productSyncOptions);
+    }
+
+    @Test
+    void save_WithSuccessfulMockCtpResponse_ShouldReturnMock() {
+        final CustomObject mock = mock(CustomObject.class);
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: Here the mock var is called `mock`, but later you have `final ProductDraft productDraft = mock(ProductDraft.class);`. Pls make it consistent.

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+                .errorCallback((errorMessage, errorException) -> {
+                    errorMessages.add(errorMessage);
+                    errorExceptions.add(errorException);
+                })
+                .build();
+        service = new UnresolvedReferencesServiceImpl(productSyncOptions);
+    }
+
+    @Test
+    void save_WithSuccessfulMockCtpResponse_ShouldReturnMock() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]:
```suggestion
    void save_WithSuccessfulMockCtpResponse_ShouldSaveMock() {
```

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 4 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: > what about fetch with unsuccessful?

As we are calling this static method `QueryExecutionUtils.queryAll` and it does not possible to mock the handle the failed called directly.

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 7 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: do you mean the this particular test case?

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 8 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: yes, as we are using the static method of `QueryExecutionUtils.queryAll` which internally calls the ctpClient.execute() we need to return a `PagedQueryResult` to handle this result inside `queryAll` method.

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/UnresolvedReferencesServiceImplIT.java [hunk 13]
```diff
@@ -92,6 +93,44 @@ void save_WithoutException_createsNewCustomObject() {
...
+            .join();
+
+
+        // assertions
+        assertThat(latestResult).hasValueSatisfying(waitingToBeResolved ->
+            assertThat(waitingToBeResolved.getProductDraft()).isEqualTo(productDraft));
+
+        final CustomObjectByKeyGet<WaitingToBeResolved> customObjectByKeyGet = CustomObjectByKeyGet
+            .of(CUSTOM_OBJECT_CONTAINER_KEY, productDraft.getKey(), WaitingToBeResolved.class);
+        final CustomObject<WaitingToBeResolved> createdCustomObject = CTP_TARGET_CLIENT
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/UnresolvedReferencesServiceImplIT.java]: It is to assert the entire `WaitingToBeResolved` object has changed to `productDraftWithUnresolvedNewRefs` in the CTP project. Basically this assert whether the old value has been replaced with new value. This test case is made because we do not maintain versioned custom objects.

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/UnresolvedReferencesServiceImplIT.java [hunk 14]
```diff
@@ -92,6 +93,44 @@ void save_WithoutException_createsNewCustomObject() {
...
+        final WaitingToBeResolved productDraftWithUnresolvedRefs =
+                new WaitingToBeResolved(productDraft, asSet("foo", "bar"));
+
+        unresolvedReferencesService
+                .save(productDraftWithUnresolvedRefs)
+                .toCompletableFuture()
+                .join();
+
+        final WaitingToBeResolved productDraftWithUnresolvedNewRefs =
+                new WaitingToBeResolved(productDraft, asSet("foo123", "bar123"));
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/UnresolvedReferencesServiceImplIT.java]: Because this test case is to verify the replacing the value of an existing custom object value. This `save_WithoutException_createsNewCustomObject()` particular test case assert the custom object creation for the first time.

File: src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java (hunk 2 above)
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java]: 3dea2d39e4f5b8ddd76371f5a685e88647732b0a

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 3 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: 22ce5b06a4ecb648f8606fd083faa2a6d8983714

File: src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java (hunk 1 above)
Comment (JudeNiroshan) [reply src/main/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImpl.java]: 1e2a4526c537a0e4abe2a909bfa0b007a23ec7f5

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 4 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: cc8958914da78b9df1fa42bb9ae1cf36db93b815

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 5 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: 7c7f6be17f49411d13eaae28ea4d03483b85e152

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 6 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: baef4a8b2da882e0e29202ebf1305d035702d94a

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 9 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: 025f8449e3edd339aa0bd73514c9514109a99517

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 10 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: b0a11c56132d37f29326ab6ff749f0b5f1531b39

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/UnresolvedReferencesServiceImplIT.java [hunk 15]
```diff
@@ -36,7 +37,7 @@
         "commercetools-sync-java.UnresolvedReferencesService.productDrafts";
 
 
-    @BeforeEach
+    @AfterEach
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/UnresolvedReferencesServiceImplIT.java]: f7bb2a9607ea2fe1a25c412b1262f002e5c41b47

File: src/test/java/com/commercetools/sync/commons/models/WaitingToBeResolvedTest.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+        // assertions
+        assertTrue(equalsResult);
+    }
+
+    @Test
+    void testEquals_withSimilarObjects() {
+        // preparation
+        when(productDraft.getKey()).thenReturn("test-product-key");
+        refSet.add("test-reference-key");
+        WaitingToBeResolved obj2 = new WaitingToBeResolved(productDraft, refSet);
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/commons/models/WaitingToBeResolvedTest.java]: 3cf9cabbcd01cc748d49870ad84f31700a52ff15

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 12 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: 87a27419aff2220080dd3e5ce20fa9f182e6fa27

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 11 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: 95578f1b9bfe5a0363c6be232839f7fd23bda287

File: src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java (hunk 7 above)
Comment (JudeNiroshan) [reply src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java]: 236b3d90326e21fc4c0e2c919cc4ffc03017fdb4
